Source/engine/render/clx_render.cpp

```diff
@@ -435,78 +397,83 @@ void RenderCl2OutlineClippedXY(const Surface &out, Point position, RenderSrcBack
 	if (position.y == -1) {
 		// Before-top line - can only draw south.
 		if (position.x <= 0) {
-			src.begin = RenderCl2OutlineRowClipped</*North=*/false, /*West=*/false, /*South=*/true, /*East=*/false, SkipColorIndexZero,
+			src.begin = RenderClxOutlineRowClipped</*North=*/false, /*West=*/false, /*South=*/true, /*East=*/false, SkipColorIndexZero,
 			    /*ClipWidth=*/true, /*CheckFirstColumn=*/true, /*CheckLastColumn=*/false>(out, position, src.begin, src.width, clipX, color, skipSize);
 		} else if (position.x + clipX.width >= out.w()) {
-			src.begin = RenderCl2OutlineRowClipped</*North=*/false, /*West=*/false, /*South=*/true, /*East=*/false, SkipColorIndexZero,
+			src.begin = RenderClxOutlineRowClipped</*North=*/false, /*West=*/false, /*South=*/true, /*East=*/false, SkipColorIndexZero,
 			    /*ClipWidth=*/true, /*CheckFirstColumn=*/false, /*CheckLastColumn=*/true>(out, position, src.begin, src.width, clipX, color, skipSize);
 		} else {
-			src.begin = RenderCl2OutlineRowClipped</*North=*/false, /*West=*/false, /*South=*/true, /*East=*/false, SkipColorIndexZero,
+			src.begin = RenderClxOutlineRowClipped</*North=*/false, /*West=*/false, /*South=*/true, /*East=*/false, SkipColorIndexZero,
 			    /*ClipWidth=*/true>(out, position, src.begin, src.width, clipX, color, skipSize);
 		}
 	}
 }
 
 template <bool SkipColorIndexZero>
-void RenderCl2Outline(const Surface &out, Point position, const uint8_t *src, std::size_t srcSize,
+void RenderClxOutline(const Surface &out, Point position, const uint8_t *src, std::size_t srcSize,
     std::size_t srcWidth, uint8_t color)
 {
 	RenderSrcBackwards srcForBackwards { src, src + srcSize, static_cast<uint_fast16_t>(srcWidth) };
 	if (position.x > 0 && position.x + static_cast<int>(srcWidth) < static_cast<int>(out.w())) {
-		RenderCl2OutlineClippedY<SkipColorIndexZero>(out, position, srcForBackwards, color);
+		RenderClxOutlineClippedY<SkipColorIndexZero>(out, position, srcForBackwards, color);
 	} else {
-		RenderCl2OutlineClippedXY<SkipColorIndexZero>(out, position, srcForBackwards, color);
+		RenderClxOutlineClippedXY<SkipColorIndexZero>(out, position, srcForBackwards, color);
 	}
 }
 
-} // namespace
-
-void Cl2ApplyTrans(byte *p, const std::array<uint8_t, 256> &ttbl, int numFrames)
+void ClxApplyTrans(ClxSprite sprite, const uint8_t *trn)
 {
-	assert(p != nullptr);
-
-	for (int i = 0; i < numFrames; ++i) {
-		constexpr int FrameHeaderSize = 10;
-		int nDataSize;
-		byte *dst = CelGetFrame(p, i, &nDataSize) + FrameHeaderSize;
-		nDataSize -= FrameHeaderSize;
-		while (nDataSize > 0) {
-			auto v = static_cast<uint8_t>(*dst++);
-			nDataSize--;
-			assert(nDataSize >= 0);
-			if (!IsCl2Opaque(v))
-				continue;
-			if (IsCl2OpaqueFill(v)) {
-				nDataSize--;
-				assert(nDataSize >= 0);
-				*dst = static_cast<byte>(ttbl[static_cast<uint8_t>(*dst)]);
+	// A bit of a hack but this is the only place in the code where we need mutable sprites.

```

Comment:
🤔

Reply:
I wasn't sure whether to introduce mutable sprites just for this one function.

Currently sprites store `const uint8_t *`, we could either make them mutable throughout (`uint8_t *`) or have separate mutable classes and conversions from mutable to const.

Other than this function, nothing else needs non-const `uint8_t *`.

Reply:
Do we utilize this function during run time, or is it only for loading? Maybe we could embed it in the load routine before the sprite data is made `const`?

On a side note this very function is the main cause of crashes in 1.4.1 so fare so that makes a extra alarmed to have hacks in it.

Reply:
> Maybe we could embed it in the load routine before the sprite data is made const?

The issue is that we do not have any types to represent non-const sprite data.
Anyway, this cast is completely safe, because the data did not start off as `const`.

> On a side note this very function is the main cause of crashes in 1.4.1

Interesting, we could replace this with runtime TRN application in a follow-up PR -- may cost some FPS but will have better RAM usage.